drivers/pcie/host/pcie.c

```diff
@@ -68,11 +68,78 @@ uintptr_t pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
 	return PCIE_CONF_BAR_ADDR(addr);
 }
 
+/* The first bit is used to indicate whether the list of reserved interrupts
+ * have been initialized based on content stored in the irq_alloc linker
+ * section in ROM.
+ */
+#define IRQ_LIST_INITIALIZED 0
+
+static ATOMIC_DEFINE(irq_reserved, CONFIG_MAX_IRQ_LINES);
+
+static unsigned int irq_alloc(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(irq_reserved); i++) {
+		unsigned int fz;
+
+		while ((fz = find_lsb_set(~atomic_get(&irq_reserved[i])))) {
+			fz--; /* find_lsb_set returns index + 1 */
+
+			if (!atomic_test_and_set_bit(&irq_reserved[i], fz)) {
+				return fz + (i * sizeof(atomic_val_t) * 8);
+			}
+		}
+	}
+
+	return PCIE_CONF_INTR_IRQ_NONE;
+}
+
+static bool irq_is_reserved(unsigned int irq)
+{
+	return atomic_test_bit(irq_reserved, irq);
+}
+
+static void irq_init(void)
+{
+	extern uint8_t __irq_alloc_start[];
+	extern uint8_t __irq_alloc_end[];
+	const uint8_t *irq;
+
+	for (irq = __irq_alloc_start; irq < __irq_alloc_end; irq++) {
+		__ASSERT_NO_MSG(*irq < CONFIG_MAX_IRQ_LINES);
+		atomic_set_bit(irq_reserved, *irq);
+	}
+}
+
 unsigned int pcie_wired_irq(pcie_bdf_t bdf)
 {
-	uint32_t data = pcie_conf_read(bdf, PCIE_CONF_INTR);
+	unsigned int irq;
+	uint32_t data;
+
+	if (!atomic_test_and_set_bit(irq_reserved, IRQ_LIST_INITIALIZED)) {
+		irq_init();
+	}
+
+	data = pcie_conf_read(bdf, PCIE_CONF_INTR);
+	irq = PCIE_CONF_INTR_IRQ(data);
+
+	if (irq == PCIE_CONF_INTR_IRQ_NONE || irq >= CONFIG_MAX_IRQ_LINES ||
+	    irq_is_reserved(irq)) {
+
+		irq = irq_alloc();
+		if (irq == PCIE_CONF_INTR_IRQ_NONE) {
+			return irq;
+		}
+
+		data &= ~0xffU;
+		data |= irq;
+		pcie_conf_write(bdf, PCIE_CONF_INTR, data);
+	} else {
+		atomic_set_bit(irq_reserved, irq);
+	}
```

Comment:
I'm considering removing this else-branch. It's there to prevent conflicts if some devices have a pre-set IRQ but others don't, however I haven't seen such "mixed" platforms yet. The consequence of setting the bit here is that multiple calls to `pcie_wired_irq()` for the same device would each time allocate a new IRQ numbers, which would mess things up. This would happen e.g. if you repeatedly call `pcie show` in the PCIe shell.

Reply:
Another solution would be to split things into two APIs: one for allocating and one for getting the existing value. Or, add a parameter to `pcie_wired_irq()` to request whether we want to try allocation or do a read-only operation. Either of these will be more intrusive since they will then require changes to existing drivers.

Reply:
splitting the logic makes sense. There isn't much drivers using pcie_wired_irq (only 2 actually)

Reply:
About the split:
drivers would call pcie_allocate_irq(bdf). 
Under the hood it would get the irq, if proper would check if not allocated already, etc... etc...
So pcie_get_irq() could still be used by the PCIe shell and would not have this border effect of reallocating